apps/src/code-studio/components/ExportAllowedDialog.jsx

```diff
@@ -0,0 +1,744 @@
+  };
+
+  state = {
+    screen: 'intro'
+  };
+
+  componentDidUpdate(prevProps) {
+    const {isOpen, projectUpdatedAt} = this.props;
+    const {exportProjectUpdatedAt} = this.state;
+    if (isOpen && !prevProps.isOpen) {
+      recordExport('open');
+      if (projectUpdatedAt !== exportProjectUpdatedAt) {
+        // The project has changed since we last opened the dialog,
+        // reset our export state, so we will need to export again:
+        this.resetExportState();
+      }
+    }
+  }
+
+  sharingDisabled() {
+    return (
+      this.props.userSharingDisabled &&
+      ['applab', 'gamelab', 'weblab'].includes(this.props.appType)
+    );
+  }
+
+  resetExportState() {
+    this.setState({
+      exporting: false,
+      exportError: null,
+      expoUri: undefined,
+      expoSnackId: undefined,
+      iconUri: undefined,
+      splashImageUri: undefined,
+      showSendToPhone: false,
+      apkUri: undefined,
+      generatingApk: false,
+      apkError: null
+    });
+  }
+
+  close = () => {
+    const {expoUri} = this.state;
+    if (expoUri) {
+      // If we are re-opened, start at the platform page:
+      this.setState({screen: 'platform'});
+    } else {
+      // If we don't haven't succesfully exported, then clear all export
+      // state so we will start again fresh the next time:
+      this.resetExportState();
+    }
+    recordExport('close');
+    this.props.onClose();
+  };
+
+  onInputSelect = ({target}) => {
+    target.select();
+  };
+
+  showSendToPhone = () => {
+    this.setState({
+      showSendToPhone: true
+    });
+  };
+
+  onInstallExpoIOS = () => {
+    window.open(
+      'https://itunes.apple.com/app/apple-store/id982107779',
+      '_blank'
+    );
+  };
+
+  onInstallExpoAndroid = () => {
+    window.open(
+      'https://play.google.com/store/apps/details?id=host.exp.exponent&referrer=www',
+      '_blank'
+    );
+  };
+
+  async publishExpoExport() {
+    const {expoUri} = this.state;
+    if (expoUri) {
+      // We have already have exported to Expo
+      return;
+    }
+    const {exportApp, projectUpdatedAt} = this.props;
+    this.setState({
+      exporting: true,
+      exportProjectUpdatedAt: projectUpdatedAt
+    });
+    try {
+      const {expoUri, expoSnackId, iconUri, splashImageUri} = await exportApp({
+        mode: 'expoPublish'
+      });
+      this.setState({
+        exporting: false,
+        exportError: null,
+        expoUri,
+        expoSnackId,
+        iconUri,
+        splashImageUri
+      });
+    } catch (e) {
+      this.setState({
+        exporting: false,
+        exportProjectUpdatedAt: null,
+        expoUri: null,
+        expoSnackId: null,
+        exportError:
+          'Failed to publish project to Expo. Please try again later.'
+      });
+    }
+  }
+
+  async publishAndGenerateApk() {
+    if (this.state.apkUri) {
+      // We have already have generated an APK
+      return;
+    }
+
+    await this.publishExpoExport();
```

Comment:
Is it weird, from a student perspective, that generating the APK here could result in us showing this Expo error?

> Failed to publish project to Expo. Please try again later.

There's a _slight_ action-at-a-distance pattern here by depending on the state changes caused by `publishExpoExport`. I also suspect that this `await` is likely, but not guaranteed, to wait until the async `setState` calls in `publishExpoExport` complete.

I might call `exportApp` directly from `publishAndGenerateApk`, trading a little duplication for more control over what we do with the Expo results.

Reply:
This is really good feedback. My solution is a little different from what you proposed, because I know that for the iOS case, I need to do `publishExpoExport()`, but not the rest of the steps in `publishAndGenerateApk()`, so I want to keep them separate.

I've changed `publishExpoExport()` to return an object (essentially the `exportResult` or a similarly shaped object for the "already exported" or failure cases). This way we don't rely on the async `setState()` as a method to pass the results in this chained async scenario.

Reply:
That sounds great. :+1: